java/src/main/native/include/jni_utils.hpp

```diff
@@ -0,0 +1,1067 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <string>
+#include <utility>
+
+#include <jni.h>
+
+#include <nvstrings/NVCategory.h>
+#include <nvstrings/NVStrings.h>
+#include <rmm/rmm.h>
+
+#include "cudf/cudf.h"
+#include "cudf/table.hpp"
+#include "utilities/column_utils.hpp"
+#include "utilities/error_utils.hpp"
+
+namespace cudf {
+namespace jni {
+
+/**
+ * @brief indicates that a JNI error of some kind was thrown and the main
+ * function should return.
+ */
+class jni_exception : public std::runtime_error {
+public:
+  jni_exception(char const *const message) : std::runtime_error(message) {}
+  jni_exception(std::string const &message) : std::runtime_error(message) {}
+};
+
+/**
+ * @brief throw a java exception and a C++ one for flow control.
+ */
+inline void throw_java_exception(JNIEnv *const env, const char *clazz_name, const char *message) {
```

Comment:
`clazz_name`?

Reply:
It is a java thing when using reflection.  Happy to change it.